pinot-core/src/main/java/org/apache/pinot/core/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java

```diff
@@ -181,33 +188,51 @@ public void seal()
     }
 
     // Create bitmaps from inverted index buffers and serialize them to file
-    try (DataOutputStream offsetDataStream = new DataOutputStream(
-        new BufferedOutputStream(new FileOutputStream(_invertedIndexFile)));
-        FileOutputStream bitmapFileStream = new FileOutputStream(_invertedIndexFile);
-        DataOutputStream bitmapDataStream = new DataOutputStream(new BufferedOutputStream(bitmapFileStream))) {
-      int bitmapOffset = (_cardinality + 1) * Integer.BYTES;
-      offsetDataStream.writeInt(bitmapOffset);
-      bitmapFileStream.getChannel().position(bitmapOffset);
-
+    ByteBuffer offsetBuffer = null;
+    ByteBuffer bitmapBuffer = null;
+    try (FileChannel channel = new RandomAccessFile(_invertedIndexFile, "rw").getChannel()) {
+      // map the offsets buffer
+      final int startOfBitmaps = (_cardinality + 1) * Integer.BYTES;
+      int bitmapOffset = startOfBitmaps;
+      offsetBuffer = channel.map(FileChannel.MapMode.READ_WRITE, 0, bitmapOffset).order(LITTLE_ENDIAN);
+      offsetBuffer.putInt(reverseBytes(bitmapOffset));
+      RoaringBitmap[] bitmaps = new RoaringBitmap[_cardinality];
```

Comment:
Do `RoaringBitmap` and `MutableRoaringBitmap` serialize to the same bytes?

Reply:
For off-heap creator, we don't want to keep all bitmaps on heap. We should try to create and serialize the bitmaps one by one. Not sure about the cost if we map a buffer per bitmap

Reply:
> Do RoaringBitmap and MutableRoaringBitmap serialize to the same bytes?

Yes, they do.

> For off-heap creator, we don't want to keep all bitmaps on heap. We should try to create and serialize the bitmaps one by one. Not sure about the cost if we map a buffer per bitmap

I can see the incentive not to allocate so much on heap, but the alternative is mapping many times (and it's a relatively costly operation) or mapping too many bytes and truncating the channel. I have done the latter in the past to get round this, but there may be downsides I'm unaware of. I've updated to use this strategy and allocation pressure is indeed reduced.